server/middleware.ts

```diff
@@ -0,0 +1,201 @@
+		}
+	}),
+	"limits": {
+		"fileSize": MAX_FILE_SIZE,
+		"files": getMaxFileUploads()
+	}
+});
+
+function getMaxFileUploads(): number {
+	// Can't use validateSchema() because this function needs to run synchronously to export uploadHandler before it gets used
+	let questionBranches: QuestionBranches = JSON.parse(fs.readFileSync(config.questionsLocation, "utf8"));
+	let questions = questionBranches.map(branch => branch.questions);
+	let fileUploadsPerBranch: number[] = questions.map(branch => {
+		return branch.reduce((prev, current) => {
+			if (current.type === "file") {
+				return prev + 1;
+			}
+			return prev;
+		}, 0);
+	});
+	return Math.max(...fileUploadsPerBranch);
+}
+
+export function isUserOrAdmin(request: express.Request, response: express.Response, next: express.NextFunction) {
+	let user = request.user as IUser;
+	if (!request.isAuthenticated()) {
+		response.status(401).json({
+			"error": "You must log in to access this endpoint"
+		});
+	}
+	else if (user._id.toString() !== request.params.id && !user.admin) {
+		response.status(403).json({
+			"error": "You are not permitted to access this endpoint"
+		});
+	}
+	else {
+		next();
+	}
+}
+
+export function isAdmin(request: express.Request, response: express.Response, next: express.NextFunction) {
+	let user = request.user as IUser;
+	if (!request.isAuthenticated()) {
+		response.status(401).json({
+			"error": "You must log in to access this endpoint"
+		});
+	}
+	else if (!user.admin) {
+		response.status(403).json({
+			"error": "You are not permitted to access this endpoint"
+		});
+	}
+	else {
+		next();
+	}
+}
+
+// For API endpoints
+export function authenticateWithReject(request: express.Request, response: express.Response, next: express.NextFunction) {
+	if (!request.isAuthenticated()) {
+		response.status(401).json({
+			"error": "You must log in to access this endpoint"
+		});
+	}
+	else {
+		next();
+	}
+}
+// For directly user facing endpoints
+export function authenticateWithRedirect(request: express.Request, response: express.Response, next: express.NextFunction) {
+	if (!request.isAuthenticated()) {
+		response.redirect("/login");
+	}
+	else {
+		next();
+	}
+}
+import * as Handlebars from "handlebars";
+import { ICommonTemplate } from "./schema";
+export enum ApplicationType {
+	Application, Confirmation
+}
+export async function timeLimited(request: express.Request, response: express.Response, next: express.NextFunction) {
+	let requestType: ApplicationType = request.url.match(/^\/apply/) ? ApplicationType.Application : ApplicationType.Confirmation;
+
+	let openBranches: (ApplicationBranch | ConfirmationBranch)[];
+	if (requestType === ApplicationType.Application) {
+		openBranches = await BranchConfig.getOpenBranches<ApplicationBranch>("Application");
+	}
+	else {
+		openBranches = await getOpenConfirmationBranches(request.user as IUser);
+	}
+
+	if (openBranches.length > 0) {
+		next();
+		return;
+	}
+
+	// TODO reimplement open and close times
```

Comment:
remove comments and/or complete this todo?

Reply:
I think we should document this in an issue and fix in future PR